### PR TITLE
kv: remove broken attempt to reject lease acquisitions on draining nodes

### PR DESCRIPTION
**Pull request by nvb**
Related to #83261.

This commit removes "protection" that avoided lease acquisitions on draining nodes. This protection had already been effectively disabled by acc1ad1, which allowed Raft leaders to bypass the check. As the comment here (added in 5ffaa9e) explained, Raft followers are already unable to acquire the lease. If leaders bypass the check and follower (and candidates) don't need it, the check is useless, so we remove it.

The commit also removes `TestReplicaDrainLease`, which was supposed to test this behavior. We remove the test not because it started failing after the change, but because it did not. It must not have been testing anything real anymore after acc1ad1.

Release justification: low risk change related to release blocker.

Release note: None

### CODE REVIEW

**Comment by aayushshah15** [comment]
*[`pkg/kv/kvserver/replica_range_lease.go` line 824 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87885#-NCLiheMDMscdLFfnZxH:-NCLiheMDMscdLFfnZxI:b7x8uyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f41d5209c557888da158b28a1264d8859603c1/pkg/kv/kvserver/replica_range_lease.go#L824)):*
> ```Go
> 			)
> 		}
> 		log.Info(ctx, "trying to take the lease while we're draining since we're the raft leader")
> ```

I thought that this log line was useful in cases where draining nodes spuriously re-acquire leases. Do you think we can still keep this log line around? Do you think there's a better way to have this observability?

**Review by irfansharif — APPROVED** [review]

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87885)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 824 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87885#-NCLiheMDMscdLFfnZxH:-NCQnYvC-MgElI6Dv1D7:b-q1o2gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f41d5209c557888da158b28a1264d8859603c1/pkg/kv/kvserver/replica_range_lease.go#L824)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

I thought that this log line was useful in cases where draining nodes spuriously re-acquire leases. Do you think we can still keep this log line around? Do you think there's a better way to have this observability? 
</blockquote></details>

Good point. I unified this with the below-Raft lease acquisition logging.

File: pkg/kv/kvserver/replica_proposal.go [hunk 1]
```diff
@@ -263,9 +263,12 @@ func (r *Replica) leasePostApplyLocked(
...
-		// Log lease acquisition whenever an Epoch-based lease changes hands (or verbose
-		// logging is enabled).
-		if newLease.Type() == roachpb.LeaseEpoch && leaseChangingHands || log.V(1) {
+		// Log lease acquisition whenever:
+		// 1. epoch-based lease changes hands
+		// 2. the new leaseholder is draining
+		// 3. verbose logging is enabled
+		epochLeaseChangingHands := newLease.Type() == roachpb.LeaseEpoch && leaseChangingHands
+		if epochLeaseChangingHands || r.store.IsDraining() || log.V(1) {
 			log.VEventf(ctx, 1, "new range lease %s following %s", newLease, prevLease)
```
Comment (aayushshah15) [inline pkg/kv/kvserver/replica_proposal.go:272]: Should this be a log.Infof instead?

Comment (nvanbenschoten) [reply pkg/kv/kvserver/replica_proposal.go]: Good catch. This logic was slowly adapted by https://github.com/cockroachdb/cockroach/commit/c3dd4724de85726fd4fac4ba02828f48622295e6 and https://github.com/cockroachdb/cockroach/commit/c3dd4724de85726fd4fac4ba02828f48622295e6. I think the change I made strikes the desired balance of verbosity, but PTAL.

**Review by aayushshah15 — APPROVED** [review]

**Comment by nvanbenschoten** [comment]
TFTR!

bors r+

**Pull request merged**